File: toplas2021/benchmarks/seahorn/03-prusti/prusti-4-pass-paper_examples-borrows_align.c
#include "seahorn/seahorn.h"
#include <stdlib.h>

typedef struct Point {
  int x, y;
} Point;
typedef struct Pair {
  Point *_0, *_1;
} Pair;

extern Point nd();

void shift_x(Point *p, int s) { p->x = p->x + s; }

int main() {
  Pair segm = {(Point *)malloc(sizeof(Point)), (Point *)malloc(sizeof(Point))};
  *segm._0 = nd();
  *segm._1 = nd();
  Point **endBox = &segm._1;
  Point *endPt = &**endBox;
  int end_x = endPt->x;
  shift_x(endPt, segm._0->x - end_x);
  sassert(segm._0->x == segm._1->x);
  return 0;
}

File: toplas2021/benchmarks/rust-horn/03-prusti/prusti-4-pass-paper_examples-borrows_align.rs
fn rand<T>() -> T { unimplemented!() }

struct Point {
  x: i32,
  y: i32,
}

fn shift_x(p: &mut Point, s: i32) { p.x = p.x + s }

fn main() {
  let mut segm: (Box<Point>, Box<Point>) = rand();
  let endBox = &mut segm.1;
  let endPt = &mut **endBox;
  // segm.1, endBox are not usable here
  let end_x = (*endPt).x;
  shift_x(endPt, (*segm.0).x - end_x);
  // borrows expire: segm.1 usable again
  assert!((*segm.0).x == (*segm.1).x);
}
